Pydra paper ania (#1)
djarecka/scipy_proceedings: papers/pydra/paper.rst

:author: Dorota Jarecka
:email: djarecka@gmail.com
:institution: Massachusetts Institute of Technology, Cambridge, MA, USA

:author: Mathias Goncalves
:email: mathiasg@stanford.edu
:institution: Stanford University, Stanford, CA, USA
:institution: Massachusetts Institute of Technology, Cambridge, MA, USA

:author: Christopher J. Markiewicz
:email:  todo
:institution: Stanford University, Stanford, CA, USA

:author: Oscar Esteban
:email: todo
:institution: Stanford University, Stanford, CA, USA

:author: Nicole Lo
:email: todo
:institution: Massachusetts Institute of Technology, Cambridge, MA, USA

:author: Jakub Kaczmarzyk
:email: todo
:institution: Stony Brook University School of Medicine, Stony Brook, NY, USA
:institution: Massachusetts Institute of Technology, Cambridge, MA, USA

:author: Satrajit Ghosh
:email: satra@mit.edu
:institution: Massachusetts Institute of Technology, Cambridge, MA, USA

:bibliography: mybib

--------------------------------------------------------------------------

> Pydra - a flexible and lightweight dataflow engine for scientific analyses

Reviewer: scientific analysis?

Author: move on, title stays without changes ;-)

--------------------------------------------------------------------------

.. class:: abstract

This paper presents a new lightweight dataflow engine written
in Python: Pydra. TODO

.. class:: keywords

   dataflow engine, scientific workflows, reproducibility

Introduction
------------

> Scientific workflows often require sophisticated analyses that encompass

Reviewer: ```suggestion
Scientific workflows often require sophisticated analysis that encompass
```

a large collection of algorithms.

> The algorithms, that were originally not necessarily designed to work together,

Reviewer: ```suggestion
Often times the algorithms were not originally designed to work together,
```

> often written by different authors in different times.

Reviewer: ```suggestion
and were written by different authors.
```

Some may be written in Python, while others might require calling external programs.

> It is a common practice to create semi-manual workflows that require scientist

Reviewer: ```suggestion
It is a common practice to create semi-manual workflows that require the scientists
```

> interaction to handle files and partial results between algorithms and external tools.

Reviewer: ```suggestion
to handle the files and interact with partial results from algorithms and external tools.
```

This approach is conceptually simple and easy to implement, but the resulting workflow
is often time consuming, error-prone and difficult to share with others.

> Consistency, reproducibility and scalability demand scientific workflows be organized

Reviewer: ```suggestion
Consistency, reproducibility and scalability demand scientific workflows to be organized
```

into fully automated pipelines.

> This was the motivation behind writting a new dataflow engine written in Python, *Pydra*,

Reviewer: ```suggestion
This was the motivation behind *Pydra* - a new dataflow engine written in Python,
```

that is presented in this paper.

> The package is a part of the second generation of the *Nipype* ecosystem

Reviewer: ```suggestion
The *Pydra* package is a part of the second generation of the *Nipype* ecosystem
```

(:cite:`Gorgolewski2011`, :cite:`Nipype`)
--- an open-source framework that provides a uniform interface to existing neuroimaging

> software and facilitates interaction between them.

Reviewer: ```suggestion
software and facilitates interaction between different software components.
```

workflows for a decade, providing a uniform interface to such neuroimaging packages
as FSL :cite:`Woolrich09`, ANTs :cite:`ANTs`, AFNI :cite:`AFNI`, FreeSurfer :cite:`FreeSurfer`
and SPM :cite:`spm_book`.
This flexibility has made it an ideal basis for popular preprocessing tools,
such as fMRIPrep :cite:`fMRIPrep` and C-PAC :cite:`C-PAC`.
The second generation of *Nipype* ecosystem is meant to provide additional flexibility
and is being developed with reproducibility, ease of use, and scalability in mind.
*Pydra* itself is a standalone project and is designed as a general-purpose dataflow engine
to support any scientific domain.

The goal of Pydra is to provide a lightweight dataflow engine for computational graph construction,
manipulation, and distributed execution, as well as ensuring reproducibility of scientific pipelines.
In Pydra, a dataflow is represented as a directed acyclic graph, where each node represents a 
Python function, execution of an external tool, or another reusable dataflow.
The combination of several key features makes Pydra a customizable and powerful dataflow engine:

* **Composable dataflows:** Any node of a dataflow graph can be another dataflow,

>   allowing for nested dataflows to arbitrary depths and encourages creating reusable dataflows.

Reviewer: ```suggestion
  allowing for nested dataflows of arbitrary depths and encouraging creating reusable dataflows.
```

* **Flexible semantics for creating nested loops over input sets:**

>   Any *Task* or dataflow can be run over input parameter sets and the outputs can be recombined

Reviewer: ```suggestion
  Any Task or dataflow can be run over input parameter sets and the outputs can be recombined
```

Reviewer: Czemu Task z dużej a dataflow z małej?

Author: za tym jest pewna mysl, byc moze nie najlpesza, ale jest... chcialam uzywac z duzej tych obiektow bardziej pydrowych. pozniej to jeszcze przemysle

  (similar concept to Map-Reduce model :cite:`Map-Reduce`, but Pydra extends this to graphs with nested dataflows).

* **A content-addressable global cache:** Hash values are computed for each graph and each Task.

>   Supports reuse of previously computed and stored dataflows and Tasks.

Reviewer: ```suggestion
  This supports reusing of previously computed and stored dataflows and Tasks.
```

> * **Can integrate Python functions and external (shell) commands:** Pydra can

Reviewer: ```suggestion
* **Access to Python functions and external (shell) commands:** Pydra can
```

Author: zmienie chyba na support

*Pydra* is intended to help scientific workflows which rely on significant file-based operations and
which evaluate outcomes of complex dataflows over a hyper-space of parameters.
It is important to note, that *Pydra* is not a framework for writing efficient scientific algorithms
or for use in applications where caching and distributed execution are not necessary.
Since *Pydra* relies on a filesystem cache at present it is also not
designed for dataflows that need to operate purely in memory. 

The next section will describe the *Pydra* architecture --- main package classes
and interactions between them. The *Key Features* section focuses on a set of features whose
combination distinguishes *Pydra* from other dataflow engines. The paper concludes with a set
of applied examples demonstrating the power and utility of *Pydra*.

Architecture
------------
*Pydra* architecture has three core components: *Task*, *Submitter* and *Worker*.
*Tasks* form the basic building blocks of the dataflow, while *Submitter*
orchestrates the dataflow execution model.
Different types of *Workers* allow *Pydra* to execute the task
on different compute architectures.
Fig. :ref:`classes`

> shows the Class hierarchy and links between them in the present Pydra

Reviewer: ```suggestion
shows the Class hierarchy in the present Pydra
```

> architecture. It was designed this way to decouple and allow *Workers* to

Reviewer: ```suggestion
architecture. 
```

> operate.  In order to describe *Pydra*'s most notable features in the next

Reviewer: ```suggestion
In order to describe *Pydra*'s most notable features in the next
```

> section, we briefly describe the role and function of each of these classes.

Reviewer: ```suggestion
section, we briefly describe the role of each of these classes.
```

* ``FunctionTask`` is a *Task* that executes Python functions. Most Python functions
  declared in an existing library, package, or interactively in a terminal can
  be converted to a ``FunctionTask`` by using *Pydra* decorator - ``mark.task``.

  .. code-block:: python

     import numpy as np
     from pydra import mark
     fft = mark.annotate({'a': np.ndarray,
                      'return': float})(np.fft.fft)
     fft_task = mark.task(fft)()
     result = fft_task(a=np.random.rand(512))

  `fft_task` is now a *Pydra* task and result will contain a *Pydra*'s ``Result`` object.
  In addition, the user can use Python's function annotation or another *Pydra*
  decorator |---| ``mark.annotate`` in order to specify the output. In the
  following example, we decorate an arbitrary Python function to create named
  outputs:

  .. code-block:: python

     @mark.task
     @mark.annotate(
         {"return": {"mean": float, "std": float}}
     )
     def mean_dev(my_data):
         import statistics as st
         return st.mean(my_data), st.stdev(my_data)

     result = mean_dev(my_data=[...])()

  When the task is executed `result.output` will contain two attributes: `mean`

>   and `std`. These named outputs allowing passing different outputs to

Reviewer: ```suggestion
  and `std`. Named attributes facilitate passing different outputs to
```

  different downstream nodes in a dataflow.

* ``ShellCommandTask`` is a *Task* used to run shell commands and executables.
  It can be used with a simple command without any arguments, or with specific
  set of arguments and flags, e.g.:

  .. code-block:: python

     ShellCommandTask(executable="pwd")

     ShellCommandTask(executable="ls", args="my_dir")

  The *Task* can accommodate more complex shell commands by allowing the user to
  customize inputs and outputs of the commands.
  One can generate an input
  specification to specify names of inputs, positions in the command, types of
  the inputs, and other metadata.
  As a specific example, FSL's BET command (Brain
  Extraction Tool) can be called on the command line as:

  .. code-block:: python

    bet input_file output_file -m

  Each of these inputs can be augmented as a named argument to the

>   ``ShellCommandTask``. As shown next, even an output is specified by constructing

Reviewer: Tego ostatniego zdania nie rozumiem.

Author: ta, to satra mi zmienil, pomysle nad nim

  .. code-block:: python

    bet_input_spec = SpecInfo(
        name="Input",
        fields=[
        ( "in_file", File,
          { "help_string": "input file ...",
            "position": 1,
            "mandatory": True } ),
        ( "out_file", str,
          { "help_string": "name of output ...",
            "position": 2,
            "output_file_template":
                              "{in_file}_br" } ),
        ( "mask", bool,
          { "help_string": "create binary mask",
            "argstr": "-m", } ) ],
        bases=(ShellSpec,) )

    ShellCommandTask(executable="bet",
                     input_spec=bet_input_spec)

  Outputs can also be specified separately using a similar output specification.

* ``ContainerTask`` class is a child class of ``ShellCommandTask`` and serves as
  a parent class for ``DockerTask`` and ``SingularityTask``. Both *Container Tasks*
  run shell commands or executables within containers with specific user defined
  environments using *Docker* :cite:`docker` and *Singularity* :cite:`singularity` software respectively.
  This might be extremely useful for users and projects that require environment
  encapsulation and sharing.

>   Using container technologies helps improve scientific

Reviewer: Ja bym to ostatnie zdanie usunęła. Prawda, ale trochę nie na temat

inputs should be split into parameter sets, and optionally combined after
the *Task* execution, the user can set *splitter* and *combiner* attributes of the
``State`` class. These attributes can be set by calling ``split`` and ``combine``
methods in the *Task* class. Here we provide a simple `Map-Reduce` example:

.. code-block:: python

  task_with_state =
        add2(x=[1, 5]).split("x").combine("x")

In this example, the ``State`` class is responsible for creating a list of two
separate inputs, *[{x: 1}, {x:5}]*, each run of the *Task* should get one
element from the list.
The results are grouped back when returning the result from the *Task*.
While this example
illustrates mapping and grouping of results over a single parameter, *Pydra*
extends this to arbitrary combinations of input fields and downstream grouping
over nested dataflows. Details of how splitters and combiners power *Pydra*'s
scalable dataflows are described later.

Submitter
=========

The ``Submitter`` class is responsible for unpacking *Workflows* and single
*Tasks* with or without ``State`` into standalone stateless jobs that are then

> executed by *Workers*. When the *runnable* is a *Workflow*, the *Submitter* is

Reviewer: Pewnie mi umknęło z wcześniej ale nie wiem co to jest runnable

When  ``ConcurrentFuturesWorker`` is created, ``ProcessPoolExecutor`` is used
to create a "pool" for adding the runnables. ``SlurmWorker`` creates an`sbatch`
submission script in order to execute the task, and ``DaskWorker`` make use of
Dask's ``Client`` class and its ``submit`` method. All workers use
*async functions* from *AsyncIO* in order to handle asynchronous processes. All
*Workers* rely on a `load_and_run` function to execute each job from its pickled
state.

Key Features
------------

In this section, features of *Pydra* that exemplify its utility for scientific
dataflows are presented. Individually, some of these features are present in the
numerous workflow packages that exist, but Pydra is the only software that brings
them together using a very lighweight codebase. The combination of the following
features makes *Pydra* a powerful tool in scientific computation.

Nested and Hashed Workflows
===========================

Scientific dataflows typically involve significant refinement and extensions as
science and instrumentation evolves. *Pydra* was designed to provide an easy way
of creating scientific dataflows that range from simple linear steps to complex

> nested graphs. It also enables reproducibility and reduced cost of dataflow

Reviewer: Ja bym całe zdanie od "*Pydra* was designed to... " usunęła stąd. Rozprasza od technicznego aspektu o którym opowiadasz. Takie zdania albo do wstępu albo do zakończenia juz bym dawała.

> nested graphs. It also enables reproducibility and reduced cost of dataflow

Reviewer: ```suggestion
*Pydra* reduces the cost of dataflow
```

maintenance through flexible reuse of already existing functions and *Workflows*
in new applications. The ``Workflow`` class inherits from ``TaskBase`` class

> and can be treated by users as any other *Task* and can itself be added as a node

Reviewer: ```suggestion
and can be treated by users as any other *Task*. As a result ``Workflow`` itself can be added as a node
```

> in a *Workflow*. This provides an easy way of creating nested workflows of

Reviewer: ```suggestion
in another *Workflow*. This provides an easy way of creating nested workflows of
```

would typically involve significant overhead for data management and use of
multiple nested loops. In Pydra, this is controlled by setting specific ``State``
related attributes through *Task* methods. In order to set input splitting
(or mapping), *Pydra* requires setting up a *splitter*. This is done using
*Task*'s ``split`` method. The simplest example would be a *Task* that has one
field *x* in the input, and therefore there is only one way of splitting its input.
Assuming that the user provides a list as a value of *x*, *Pydra* splits
the list, so each copy of the *Task* will get one element of the list:

.. math::

   \textcolor{red}{\mathnormal{S} = x}: x=[x_1, x_2, ..., x_n] \longmapsto x=x_1, x=x_2, ..., x=x_n

That is also represented in Fig. :ref:`ndspl1`, where *x=[1, 2, 3]* as an example.

.. figure:: nd_spl_1-crop.pdf
   :figclass: h!
   :scale: 100%

   Diagram representing a Task with one input and a simple splitter. The white
   node represents an original Task with x=[1,2,3], as an input. The coloured
   nodes represent copies of the original Task after splitting the input, these
   are the runnables that are executed by Workers.
   :label:`ndspl1`

> 

Reviewer: Nie bardzo wiem po co jest linijka z out=5, out=3 out=4.

Reviewer: I tak samo w kolejnych schematach.

Reviewer: ok - doszłam do Fig.6 i juz widzę po co. Może warto zdefiniować gdzieś co robi operator S? Żeby się głupi czytelnik nie zastanawiał czemu S(1) = 5 ;)

Author: glupi czytelnik zawsze sie znajdzie ;-)

combining the output. Taking as an example *Task* represented in Fig. :ref:`ndspl3`,
it might be useful to combine all the outputs for one specific value of *a* and
all the values of *b*. The combined output is a two dimensional list, each
inner element for each value of *a*. This is written as follows:

.. math::
   :type: eqnarray

   \textcolor{red}{\mathnormal{C} = y} &:& out(x_1, y1), out(x_1, y2), ...out(x_n, y_m) \\
    &\longmapsto& [[out(x_1, y_1), ..., out(x_1, y_m)], \\
    && ..., \\
    && [out(x_n, y_1), ..., out(x_n, y_m)]]

And is represented in Fig. :ref:`ndspl3comb1` (todo: should probably change a,b to x,y)

.. figure:: nd_spl_3_comb1-crop.pdf
   :figclass: h!
   :scale: 75%

   Diagram representing a Task with two input fields, an outer splitter and a
   combiner. The Tasks are run in exactly the same way as previously, but at the
   end the values of output for all values of *b* are combined together. The
   symbol convention is described in :ref:`ndspl1`.
   :label:`ndspl3comb1`

However, for the diagram from :ref:`ndspl3`, it might be also useful to combine
all values of *a* for specific values of *b*. One may also want to combine all

> the values together. This can be achieve by providing a list of fields,

Reviewer: ```suggestion
the values together. This can be achieved by providing a list of fields,
```

*[a, b]* to the combiner. When a full combiner is set, i.e. all the fields from
splitter are also in the combiner, the output is a one dimensional list:

.. math::
   :type: eqnarray

   \textcolor{red}{\mathnormal{C} = [x, y]} : out(x_1, y1), ...out(x_n, y_m)
    \longmapsto [out(x_1, y_1), ..., out(x_n, y_m)]

> And is represented in Fig. :ref:`ndspl3comb3` (todo: should probably change a,b to x,y)

Reviewer: todo!

any *Task* or *Workflow*, *Pydra* checks all the directories that are either in
``cache_dir`` or ``cache_locations``, and if the specific checksum is found,
then the results are reloaded without running the specific *Task*. It is important
to emphasize that without a cache, every element of a nested *Workflow* would be
re-executed. Using `Global Cache` can reduce significant execution time when
the same operations on the same data are repeated. This is also true for *Tasks*
with *State*. If the number of input elements is expanded, the previously cached
results can be reused without recomputation. For scientific workflows, where
many tasks take significant computational resources, this can speed up reruns.

Applications and Examples
-------------------------

In this section, we highlight *Pydra* through two examples. The first example
is an intuitive scientific Python example to demonstrate the power of *Pydra*'s
splitter and combiner. The second example extends this demonstration with a more
practical machine learning model comparison workflow leveraging scikit-learn.

Example 1: Perform Sine Function Approximation
==============================================

> This example illustrates the flexibility of *Pydra*'s splitters and combiners.

Reviewer: powtarza sie z poprzednim paragrafem

We do not expect Pydra to be used to write algorithms like this. The exemplary

> workflow will calculate the approximated values of Sine function for various

Reviewer: sine ?

Author: tez bylam zniesmaczona... ale tak, nie uzywaja sinus, ale sine..

to emphasize that without a cache, every element of a nested *Workflow* would be
re-executed. Using `Global Cache` can significantly reduce execution time when
the same operations on the same data are repeated. This is also true for *Tasks*
with *State*. If the number of input elements is expanded, the previously cached
results can be reused without recomputation. For scientific workflows, where
many tasks take significant computational resources, this can drastically speed up reruns.

Applications and Examples
-------------------------

In this section, we highlight *Pydra* through two examples. The first example
is an intuitive scientific Python example to demonstrate the power of *Pydra*'s
splitter and combiner. The second example extends this demonstration with a more
practical machine learning model comparison workflow leveraging scikit-learn.

Example 1: Sine Function Approximation
======================================

This example illustrates the flexibility of the *Pydra*'s *splitters* and *combiners*
(the example is not meant to convince scientist to use *Pydra* to write algorithms like this).

> The exemplary workflow will calculate the approximated values of *Sine* function for various

Reviewer: sine?
